python/tests/test_tables.py

```diff
@@ -851,6 +851,67 @@ def test_round_trip_set_columns(self):
                 assert table[j].metadata == metadata_column[j]
                 assert table[j + num_rows].metadata == metadata_column[j]
 
+    def test_set_null_metadata(self):
+        for codec in ["struct", "json"]:
+            table = self.table_class()
+            table.metadata_schema = metadata.MetadataSchema(
+                {
+                    "codec": f"{codec}",
+                    "title": "Example Metadata",
+                    "type": ["object", "null"],
+                    "properties": {
+                        "a": {"type": "number", "binaryFormat": "i"},
+                    },
+                    "required": ["a"],
+                    "additionalProperties": False,
+                },
+            )
+            metadata_list = [{"a": 4}, None, {"a": 5}]
+            for md in metadata_list:
+                table.add_row(
+                    **{
+                        **self.input_data_for_add_row(),
+                        "metadata": md,
+                    }
+                )
+            assert table.num_rows == len(metadata_list)
+            for md, row in zip(metadata_list, table):
+                assert md == row.metadata
+
+    def test_set_with_optional_properties(self):
+        for codec in ["json", "struct"]:
+            print("codec", codec)
+            table = self.table_class()
+            table.metadata_schema = metadata.MetadataSchema(
+                {
+                    "codec": f"{codec}",
+                    "title": "Example Metadata",
+                    "type": ["object", "null"],
+                    "properties": {
+                        "a": {"type": "number", "binaryFormat": "i", "default": 0},
+                    },
+                    "required": [],
+                    "additionalProperties": False,
+                },
+            )
+            metadata_list = [{"a": 4}, None, {"a": 5}, {}]
+            for md in metadata_list:
+                table.add_row(
+                    **{
+                        **self.input_data_for_add_row(),
+                        "metadata": md,
+                    }
+                )
+            assert table.num_rows == len(metadata_list)
+            for md, row in zip(metadata_list, table):
+                if isinstance(md, dict):
+                    # struct inserts defaults, json does not
```

Comment:
This was surprising. Does the JSON codec do anything with defaults?

Reply:
It's in work here: https://github.com/tskit-dev/tskit/pull/1149